src/ftl/ast/ast.js

```diff
@@ -18,87 +18,108 @@ class Resource extends NodeList {
   }
 }
 
-class Section extends NodeList {
-  constructor(key, body = [], comment = null) {
-    super(body, comment);
-    this.type = 'Section';
-    this.key = key;
-  }
-}
-
 class Entry extends Node {
   constructor() {
     super();
     this.type = 'Entry';
   }
 }
 
-class Identifier extends Node {
-  constructor(name) {
+class Message extends Entry {
+  constructor(id, value = null, attributes = null, comment = null) {
     super();
-    this.type = 'Identifier';
-    this.name = name;
+    this.type = 'Entity';
+    this.id = id;
+    this.value = value;
+    this.attributes = attributes;
+    this.comment = comment;
   }
 }
 
 class Pattern extends Node {
-  constructor(source, elements, quoted = false) {
+  constructor(elements, quoted = false) {
     super();
     this.type = 'Pattern';
-    this.source = source;
-    this.elements = elements;
+    this.elements = elements
     this.quoted = quoted;
   }
 }
 
-class Member extends Node {
-  constructor(key, value, def = false) {
+class Expression extends Node {
+  constructor() {
     super();
-    this.type = 'Member';
-    this.key = key;
-    this.value = value;
-    this.default = def;
+    this.type = 'Expression';
   }
 }
 
-class Entity extends Entry {
-  constructor(id, value = null, traits = [], comment = null) {
+class PatternExpression extends Expression {
+  constructor(pattern) {
     super();
-    this.type = 'Entity';
+    this.type = 'PatternExpression';
+    this.pattern = pattern;
+  }
+}
+
+class StringExpression extends Expression {
+  constructor(string) {
+    super();
+    this.type = 'StringExpression';
+    this.string = string;
```

Comment:
The String type isn't defined in the ASDL as it's one of ASDL's built-in types. Should we be more explicit about it and define it? I'd also vote for `StringExpression` to have a `value` field instead of `string`.